core/server/master/src/main/java/alluxio/master/throttle/SystemMonitor.java

```diff
@@ -0,0 +1,405 @@
+    // continuously, this means both previous and current status should be either STRESSED
+    // or OVERLOADED
+    if ((mPrevPitInfo.getSystemStatus() == SystemStatus.STRESSED
+        || mPrevPitInfo.getSystemStatus() == SystemStatus.OVERLOADED)
+        && (mCurrentSystemStatus == SystemStatus.STRESSED
+        || mCurrentSystemStatus == SystemStatus.OVERLOADED)
+        && mPrevFileSystemIndicators != null
+        && mCurrentFileSystemIndicators != null) {
+      mDeltaFilesystemIndicators = new FileSystemIndicator(mCurrentFileSystemIndicators);
+      mDeltaFilesystemIndicators.deltaTo(mPrevFileSystemIndicators);
+    } else {
+      mDeltaFilesystemIndicators = null;
+    }
+  }
+
+  private void collectFileSystemIndicators(long snapshotTimeMS) {
+    mPrevFileSystemIndicators = mCurrentFileSystemIndicators;
+    if (mCurrentSystemStatus == SystemStatus.IDLE
+        || mCurrentSystemStatus == SystemStatus.ACTIVE) {
+      // don't collect information in case of IDLE and ACTIVE
+      mCurrentFileSystemIndicators = null;
+      mDeltaFilesystemIndicators = null;
+      return;
+    }
+
+    FileSystemIndicator fileSystemIndicator = new FileSystemIndicator();
+    fileSystemIndicator.setPitTimeMS(snapshotTimeMS);
+
+    mCurrentFileSystemIndicators = fileSystemIndicator;
+    produceDeltaFilesystemIndicators();
+  }
+
+  private void collectServerIndicators(long snapshotTimeMS) {
+    ServerIndicator serverIndicator
+        = ServerIndicator.createFromMetrics(
+        mPrevPitInfo != null ? mPrevPitInfo.getJVMPauseTime() : 0);
+    serverIndicator.setPitTimeMS(snapshotTimeMS);
+
+    mServerIndicatorsList.add(serverIndicator);
+    if (mServerIndicatorsList.size() > mMaxNumberOfSnapshot && !mServerIndicatorsList.isEmpty()) {
+      ServerIndicator toBeRemoved = mServerIndicatorsList.get(0);
+      if (mAggregatedServerIndicators != null) {
+        mAggregatedServerIndicators.reduction(toBeRemoved);
+      }
+      mServerIndicatorsList.remove(0);
+    }
+
+    if (mAggregatedServerIndicators == null) {
+      mAggregatedServerIndicators = new ServerIndicator(serverIndicator);
+    } else {
+      mAggregatedServerIndicators.addition(serverIndicator);
+    }
+  }
+
+  /**
+   * Checks whether to escalate or deescalate, if the pit low boundary is null, not deescalated,
+   * and if the high aggregated boundary is null, not escalated.
+   *
+   * @param aggregatedServerIndicator the aggregated server indicator
+   * @param pitServerIndicator the pit server indicator
+   * @param lowBoundaryPitIndicator the low pit boundary indicator
+   * @param highBoundaryPitIndicator the high pit boundary indicator
+   * @param lowBoundaryAggregateIndicator the low aggregated boundary indicator
+   * @param highBoundaryAggregateIndicator the high aggregated boundary indicator
+   * @return Whether to escalate, deescalate or no change
+   */
+  private StatusTransition statusCheck(ServerIndicator aggregatedServerIndicator,
```

Comment:
Can we always check all indicators (no need to pass in indicators) and return the status check result e.g. SystemStatus.IDLE/ACTIVE....?  `SystemStatus statusCheck()`
and give SystemStatus a number e.g. IDLE=1, ACTIVE=2,     STRESSED = 3, OVERLOADED = 4,

by comparing the current SystemStatus  number and the newly returned SystemStatus number by `statusCheck` method, then we know the status should be escalated, unchanged, or deescalated and can be done in a uniformed way.

because it's kind of hard to know which indicator should be passed in for this method and maintain the relatively complex StatusTransition logic

Reply:
See https://github.com/Alluxio/alluxio/pull/16190/files#r991856230 
The parameters are passed according to current status.

Reply:
The status has is low boundary and up boundary. Low boundary is a point in time value, and up boundary is an aggregated value.

To check whether to deescalate is using pit, and comparing with low boundary.
To check whether to escalate is using aggregated(T-2, T-1, T), and comparing with the up boundary

Reply:
why cannot get a "target" SystemStatus based on all the different indicator check result?
why need to escalate/deescalate based on different criteria? 

https://github.com/Alluxio/alluxio/pull/16190/files#r991856230 yeah, from a pure code reader perspective, without the background design doc/knowledge, it's relative hard to understand the escalate/deescalate logics here and understand which indicator should be passed in to the method and which should not

Reply:
comments added